Keycloak 14 compatibility and more necessary features

haraldsegliens opened the pull request:
Found this keycloak configuration tool. Works very well for me. Unfortunately everything is not perfect, therefore contributed some fixes/features so that I can benefit from this tool.

1. Made the tool work with Keycloak 14.  defaultRoles -> defaultRole change had be made.
2. Realm's theme configuration
3. Required action configuration
4. Config parameter for flow executions
5. protocolMappers parameter for clientScope
6. configure any user federation not only for LDAP

Problems that still need to be solved are:

- defaultRoles -> defaultRole change make this tool not compatible with Keycloak versions that don't have defaultRole. Should this tool detect keycloak version and send appropriate messages or some versioning should be done?
- How to contribute in documentation?
- I still need to write unit tests for my changes. Will start to do that when I will have clarity on the first question.

klg71 commented:
Hey harald thanks for contributing :) ♥

I guess we can release the version with a note that its only compatible with version 13.0 upwards (thats where the defaultRoles change occured imho).
For documentation you have to install the `hugo`(https://gohugo.io/) static weg page renderer, put it in path and then you can execute the gradle task `:docsbuild:buildDocs`. That puts the rendered documentation into the `docs` folder for you to inspect.
This folder is also comitted so that github pages can render the contents of this directory.

klg71 commented on build.gradle.kts:
```diff
@@ -134,7 +134,7 @@ tasks {
         workingDir("keycloak/keycloak-$keycloakVersion/bin")
         commandLine("cmd", "/c", "add-user-keycloak.bat", "-r", "master", "-u", "admin", "-p", "admin")
-        environment("JAVA_OPTS", "--add-modules=java.se")
+        //environment("JAVA_OPTS", "--add-modules=java.se")
```
if its not necessary you might just remove it :)

haraldsegliens replied:
removed it

klg71 commented on keycloakapi/src/main/kotlin/de/klg71/keycloakmigration/keycloakapi/KeycloakClient.kt:
```diff
@@ -1,43 +1,6 @@
-import de.klg71.keycloakmigration.keycloakapi.model.UserFederation
-import de.klg71.keycloakmigration.keycloakapi.model.UserFederationMapper
-import de.klg71.keycloakmigration.keycloakapi.model.UserGroupListItem
+import de.klg71.keycloakmigration.keycloakapi.model.*
```
I personally don't like star imports but if you don't want to change them back I will change it before i release a new version.

haraldsegliens replied:
done. brought back non star imports. Forgot to disable star import usage in Intelij IDEA

klg71 commented on src/main/kotlin/de/klg71/keycloakmigration/changeControl/actions/requiredactions/UpdateRequiredActionAction.kt:
```diff
@@ -0,0 +1,63 @@
+class UpdateRequiredActionAction(
+    realm: String?,
+    private val alias: String,
+    private val providerId: String?,
```
I would prefer defining all nullable values with null as default parameter

haraldsegliens replied:
added null default value to default parameters

klg71 reviewed (changes requested):
Overall really good pull requests and thank you for contributing :) I just commented some minor issues that I would like to change.

haraldsegliens commented:
Fixed minor issues that have been raised.
Updated integration tests that verify correctness of the changes. While I was testing the changes, noticed that UpdateRealm action doesn't update defaultRole parameter, therefore removed the parameter. So probably this tool is still compatible with Keycloak 12 and less.
Also update documentation.

klg71 commented:
Did some really minor adjustments and released it with version 0.2.21 please confirm it works :) Thanks for contributing and up to the next time :)

haraldsegliens commented:
Tested 0.2.21 keycloakmigration tool. I confirm that it works.
